core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreAggregateCaseAggregations.java

```diff
@@ -282,30 +285,37 @@ private ProjectNode createPreProjection(
 
     private Map<PreAggregationKey, PreAggregation> getPreAggregations(List<CaseAggregation> aggregations, Context context)
     {
-        Set<PreAggregationKey> keys = new HashSet<>();
-        ImmutableMap.Builder<PreAggregationKey, PreAggregation> preAggregations = ImmutableMap.builder();
-        for (CaseAggregation aggregation : aggregations) {
-            PreAggregationKey preAggregationKey = new PreAggregationKey(aggregation);
-            if (keys.contains(preAggregationKey)) {
-                continue;
-            }
+        SetMultimap<PreAggregationKey, CaseAggregation> groupedAggregations = aggregations.stream()
+                .collect(toImmutableSetMultimap(PreAggregationKey::new, identity()));
+
+        return groupedAggregations.asMap().entrySet().stream().collect(toImmutableMap(
+                Map.Entry::getKey,
+                entry -> {
+                    PreAggregationKey key = entry.getKey();
+                    Set<CaseAggregation> caseAggregations = (Set<CaseAggregation>) entry.getValue();
+                    Expression preProjection = key.projection;
+
+                    // Cast pre-projection if needed to match aggregation input type.
+                    // This is because entire "CASE WHEN" expression could be wrapped in CAST.
+                    Type preProjectionType = getType(context, preProjection);
+                    Type aggregationInputType = getOnlyElement(key.getFunction().getSignature().getArgumentTypes());
+                    if (!preProjectionType.equals(aggregationInputType)) {
+                        preProjection = new Cast(preProjection, toSqlType(aggregationInputType));
+                        preProjectionType = aggregationInputType;
+                    }
 
-            // Cast pre-projection if needed to match aggregation input type.
-            // This is because entire "CASE WHEN" expression could be wrapped in CAST.
-            Expression preProjection = aggregation.getResult();
-            Type preProjectionType = getType(context, preProjection);
-            Type aggregationInputType = getOnlyElement(aggregation.getFunction().getSignature().getArgumentTypes());
-            if (!preProjectionType.equals(aggregationInputType)) {
-                preProjection = new Cast(preProjection, toSqlType(aggregationInputType));
-                preProjectionType = aggregationInputType;
-            }
+                    // Wrap the preProjection with IF to retain the conditional nature on the CASE aggregation(s) during pre-aggregation
+                    if (!(preProjection instanceof SymbolReference || preProjection instanceof Literal)) {
```

Comment:
There might be some regressions due to this, so it would be good to narrow down it a bit further, but we don't have to do this as part of this PR